### DISCUSSION TITLE
Future breaking changes in libgit2 v2.0

### DISCUSSION THREAD

**Discussion by ethomson** [libgit2/libgit2#6988] [Announcements]
We expect [libgit2 v1.9](http://github.com/libgit2/libgit2/releases/v1.9.0) to be the last release in the libgit2 v1.x release line. libgit2 v2.0 will include a number of breaking changes.

* **SHA256 support**
   SHA256 support will graduate from "experimental" to be a supported part of the library. This will include both API and ABI changes to support.

* **TLS v1.2 as a minimum**
   libgit2 will remove support for HTTPS versions prior to TLS v1.2 and will update to the "intermediate" settings [documented by Mozilla](https://wiki.mozilla.org/Security/Server_Side_TLS).

* **libtool-compatible SONAME versioning**
   Beginning with libgit2 v2.0, we will use SONAME versioning standards that align with [libtool versioning](https://www.gnu.org/software/libtool/manual/html_node/Libtool-versioning.html) requirements. Meaning, we will update the SONAME major version number on incompatible ABI updates. In addition, we will attempt to keep more ABI compatibility between minor versions of the library.

* **Removing the chromium zlib built-in**
   We will remove the ability for our cmake build system to compile a chromium zlib implementation in the "deps" directory. Users are welcome to build against chromium zlib, but should build it separately from libgit2, then link against it.

* **Removing the libssh2 embedded build system**
   Similarly, we will remove the ability for our cmake build system to compile libssh2. Users are welcome to build against libssh2, but will need to build libssh2 separately from libgit2, then link against it.

* **WinHTTP deprecation**
   WinHTTP will no longer be the default HTTP provider for Windows, and will be deprecated for behavioral compatibility across operating systems.

* **Object type cleanup**
   The `git_object_t` enum contains items like `GIT_OBJECT_OFS_DELTA`, which is _not an object type_, it's a descriptor of an item in a packfile. Similarly, the API `git_object_type_isloose` is nonsensical. The former should not exist, and the latter will be updated to talk about type validity.

**Comment by ethomson** [comment]
Note that one thing that was mentioned in https://github.com/libgit2/libgit2/pull/6191 was the idea that carrying around larger `git_oid`s that support SHA256 is wasteful for SHA1 repositories. This is, of course, true, but not something that we've addressed for v2.0.

I want to be a bit more concrete about this need and use cases before I write any code, but one can imagine something along the lines of:

```c
typedef struct git_sha1_oid {
    /** raw binary formatted id */
    unsigned char id[GIT_OID_SHA1_SIZE];
} git_sha1_oid;

GIT_EXTERN(int) git_sha1_blob_lookup(
    git_blob **blob,
    git_repository *repo,
    const git_sha1_oid *id);

...

int git_sha1_blob_lookup(git_blob **out, git_repository *repo, const git_sha1_oid *id)
{
    git_oid proxy;

    proxy.type = GIT_OID_SHA1;
    memcpy(&proxy, id->id, GIT_OID_SHA1_SIZE);

    return git_blob_lookup(out, repo, &proxy);
}
```

Similarly, for things that return or write to OIDs, we could do the opposite translation. This seems straightforward (and something we could probably even produce mechanically.) One could even imagine that you could do some dynamic loader trickery so that this "just works", but I'm not sure if that really makes sense. (It sounds complicated to me, and potentially brittle.)

There are a couple of issues here:

1. It's not clear where this code should live. One could imagine that it lives within libgit2, a sidecar library that we provide, or even the caller's code. If there's broad need for this, then we should ship it; if not, it's probably not worth us investing in.
2. More importantly, _what do we do about callbacks_? There are a handful of callbacks that give object IDs to callers. Presumably if you're using this odd API then you're all in on `git_sha1_oid` or whatever we call it. I'd want to give you `git_sha1_oid` in your callbacks as well. That's a little trickier; we could certainly build out proxies for all the things, but, meh.

Anyway. This is certainly possible, and I definitely want to give people the opportunity to save precious bytes. But without a clear need and a user who wants to help design this (or, better, help code it) then I'm going to hit pause ⏸️ here and say that we have an idea of what we would do, we're just waiting for the demand.

_In the meantime_, I intend to give people an opt-out to turn _off_ SHA256 support when they build to avoid those wasted bytes if they think that it's meaningful to them. (Basically inverting the `EXPERIMENTAL_SHA256`.)

**Comment by Uncommon** [comment]
One concern I have with SHA256 support is performance. In my code for constructing a commit history, I found that `git_oid_equal()` was a bottleneck: even when building with `GIT_EXPERIMENTAL_SHA256` turned off, the switch statement in `git_oid_size` was an extra step that could really add up in large repositories. I ended up just calling `memcmp` directly, but this of course won't work once I actually need SHA256 support.

Is there a way we can have an approach where libgit2 knows that it's a SHA256 repo to begin with, and doesn't need to constantly check?

**Reply by ethomson** [reply]
Great question. I spend too much of my time in debugging builds. It seems like taking an optimization pass would be useful. Let me spend some time on this!

**Reply by ethomson** [reply]
@Uncommon can you share what processor, OS version and compiler version you're using?

**Reply by Uncommon** [reply]
I'm working on an M1 MacBook Pro. Looking at my repo history, I made the switch to `memcmp` in April last year so that would have been macOS 14 and Xcode 15.

**Reply by ethomson** [reply]
Interesting - I think that I misunderstood. You're saying that you saw a slowdown in `git_oid_equal` _even when you're not compiling with Experimental SHA256 support_? Thanks. This is helpful.

**Reply by Uncommon** [reply]
Yes, the `switch` statement in `git_oid_size()` ended up taking a bit of time even when it only had one case. It seems like maybe that should get optimized out, but it wasn't.

**Reply by ethomson** [reply]
Thanks. I've been digging in to the optimizations that clang is doing; I naively assumed that gcc and clang optimized roughly equivalently on arm64 and also did a reasonably poor job building a microbenchmark for this in the first place. I'll investigate and will ship a v1.9.1.

**Reply by ethomson** [reply]
It looks like you made this switch here — https://github.com/Uncommon/Xit/commit/738ca6016f509acf4b9a0daf1bb9d439b65de8b8 — I'm curious if you happen to remember any more details? I ask because I'm having a bit of trouble replicating your report.

Digging a bit deeper - here: https://github.com/Uncommon/Xit/blame/e7ac31506a7666cedf2021dba00f5ed2a21e1d69/Xit/Repository/OID.swift you were using a custom wrapper over `memcpy`, as the comment indicates, our `git_oid_cmp` did a byte-for-byte comparison, which was very surprising.

It looks like here: https://github.com/Uncommon/Xit/blame/9b05de72d0bcf9da4447e566ff88474c430ddc2d/Xit/Repository/OID.swift you started using `git_oid_equal` when you upgraded to libgit2 v1.5.0. 

and here: https://github.com/Uncommon/Xit/commit/738ca6016f509acf4b9a0daf1bb9d439b65de8b8 you moved back to a `memcmp`.

When I run a few million `git_oid_equals` against main _or_ v1.5.0, I don't see a speed difference. I also notice that libgit2 v1.5.0 already used the `git_oid_size`. Also, removing the `git_oid_size` and hardcoding it does not make a difference to me. So I wonder if something else is going on.

I'm not surprised that `memcmp == 0` would be faster than a call into libgit2 for `git_oid_equal`. So I'm not sure that I can get parity to a `memcmp`; gcc in particular is good at optimizing `memcmp`. But I would like to speed it up, especially if there was a regression.

libgit2 v1.5.0 was 2.5 years ago. Any chance you were still using an x86 Mac then? I wonder if there are more obvious differences on x86 than arm. I can poke around at that hypothesis tomorrow.

**Reply by Uncommon** [reply]
I'm pretty sure I've had my M1 Mac for longer than that. What I recall is that I saw `git_oid_equal` turn up in the profiling results, so I made the switch, and then it no longer appeared as a significant factor in the profiling. I don't recall what the numbers were like, though.
